addons/viewport/src/manager/components/Panel.js

```diff
@@ -26,20 +47,42 @@ export class Panel extends Component {
     super(props, context);
     this.state = {
       viewport: defaultViewport,
+      viewports: initialViewports,
       isLandscape: false,
     };
-
-    this.props.channel.on('addon:viewport:update', this.changeViewport);
   }
 
   componentDidMount() {
     this.iframe = document.getElementById(storybookIframe);
+
+    this.props.channel.on(UPDATE_VIEWPORT_EVENT_ID, this.changeViewport);
+    this.props.channel.on(ADD_VIEWPORTS_EVENT_ID, this.addViewports);
+    this.props.channel.on(SET_VIEWPORTS_EVENT_ID, this.setViewports);
   }
 
   componentWillUnmount() {
-    this.props.channel.removeListener('addon:viewport:update', this.changeViewport);
+    this.props.channel.removeListener(UPDATE_VIEWPORT_EVENT_ID, this.changeViewport);
+    this.props.channel.removeListener(ADD_VIEWPORTS_EVENT_ID, this.addViewports);
+    this.props.channel.removeListener(SET_VIEWPORTS_EVENT_ID, this.setViewports);
   }
 
+  setViewports = viewports => {
+    const newViewports = viewportsTransformer(viewports);
+
+    this.setState({ viewports: newViewports });
+  };
+
+  addViewports = viewports => {
+    const newViewports = viewportsTransformer(viewports);
+
+    this.setState({
+      viewports: {
+        ...initialViewports,
```

Comment:
Should it be current viewports instead?

Reply:
My intention was to make `add` works by adding new viewports to the initial ones, but indeed your point makes more sense

Reply:
done